Adding design guidelines for modeling collections with state associations

duchau opened the pull request.

dulcineachau committed:
Create State-Associated-Subsets.md

Adding new guidlines for subsets.

dulcineachau committed:
Update State-Associated-Subsets.md

Fixing spaces

dulcineachau committed:
Update State-Associated-Subsets.md

Fixing json

mikepizzo reviewed (commented):
Thanks for the quick turn around on this @duchau!

See my comments in the email thread. We made some revisions to the guidance in our API Council meeting, largely around adding a memberType enum to the base type, which both makes the payloads easier to read and simplifies querying on all versus an enumerated membership.

dulcineachau committed:
Adding enum requirement

Including scope enum in base entity to allow for simpler filtering capabilities

duchau commented:
> Thanks for the quick turn around on this @duchau!
> 
> See my comments in the email thread. We made some revisions to the guidance in our API Council meeting, largely around adding a memberType enum to the base type, which both makes the payloads easier to read and simplifies querying on all versus an enumerated membership.

Addressed in latest iteration.

dulcineachau committed:
Pulling Dan's changes in

darrelmiller commented on graph/State-Associated-Subsets.md:
```diff
@@ -0,0 +1,113 @@
+## Problem
+--------
+
+Existing patterns for this either have special-cased 'strings' or have tightly coupled dependencies between two independent properties. Neither are intuitive, require reading documentation and cannot be inferred from the schema or within client libraries.
```
The problem statement should be expressing the problem that is being solved by this pattern, not discussing the disadvantages of other patterns that try to solve this problem.

darrelmiller commented on graph/State-Associated-Subsets.md:
```diff
@@ -0,0 +1,113 @@
+      <Property Name="membershipKind" Type="graph.membershipKind"/>
+    </ComplexType>
+
+    <Enum Name="membershipKind">
```
As this is duplicating the @OData.Type as a discriminator, we should be explicit as to why we think it is necessary in this case.

darrelmiller commented on graph/State-Associated-Subsets.md:
```diff
@@ -0,0 +1,113 @@
+
+
+
+Modeling state associated to a collection that may include all instances, an included subset, an excluded subset, no instances, or any combinations of the above.
```
This pattern doesn't seem to be associating any state to a collection.  It feels more like defining a set of well known predicates for the purpose of determining subsets of a collection.   The state is the predicate attached to the policy (or whatever) that wants to operate on the collection.

gareth-a-jones commented on graph/State-Associated-Subsets.md:
```diff
@@ -0,0 +1,113 @@
+## When to Use this Pattern
+------------------------
+
+This pattern is to be used when supporting two or more collection states of the following where at least one of the states is a subset variant.
```
As Darrel points out above, this really isn't a 'state'

gareth-a-jones commented on graph/State-Associated-Subsets.md:
```diff
@@ -0,0 +1,113 @@
+                "users": {
+                    "includeGuestsOrExternalUsers": {
+                        "externalTenants": {
+                            "@odata.type":"microsoft.graph.conditionalAccessAllExternalTenants",
```
I am struggling to remember from the discussion Mike led her, but wht is the value of having BOTH the odata.type annotation AND the membershipKind enum?   Isn't the odata.typ eon its own adequate?  Perhaps I am just forgetting the reason.
